### ISSUE TITLE
Honor media HTML attribute for link manifest

### ISSUE THREAD

**Issue by beaufortfrancois** [mozilla/standards-positions#500]
## Request for Mozilla Position on an Emerging Web Specification

* Specification Title: HTML spec
* Specification or proposal URL: https://html.spec.whatwg.org/multipage/links.html#link-type-manifest
* Discussion: https://github.com/whatwg/html/pull/6444

### Other information

Browsers don’t currently honor the media attribute for `link[rel="manifest"]` even though [the updated HTML specification says](https://github.com/whatwg/html/pull/6444) they should.

> When selecting a manifest, the user agent must select the most appropriate link element whose rel attribute contains the token manifest, while taking into consideration the media and lang attributes (if either is present). If there are multiple equally appropriate link elements, the user agent must use the last one declared in tree order.

Members of the Google Chrome team are thinking of honoring the link element’s "media" attribute for link[rel="manifest"] so that web developers can define multiple web app manifests based on a media query (dark and light modes for instance). The last one that matches will be picked at "install/use time". 

Note that this intent is a first step as noted at https://github.com/whatwg/html/pull/6444#issuecomment-795171477

**Comment by annevk** [comment]
I should probably have thought about this sooner, but one thing that strikes me as odd here is that these manifests are for an application and the user needs might change over time. So while I might install application during light mode time, when I want to use it at dark mode time it would seem somewhat annoying if I was somehow stuck with a light mode entry point. Similarly, if I use a tiny browser window that doesn't mean I want to be stuck with a tiny variant of the application.

@yoav raised this in https://github.com/whatwg/html/pull/6444#issuecomment-795000109, but the response to that by @marcoscaceres was contrary to what I expected. As in, I would have expected that you want to make these kind of presentation decisions later, at time of use.

**Comment by beaufortfrancois** [comment]
Note that this proposal fixes a current problem where users who use dark mode always get the light splash screen color when they open their web app.

I wonder if in the future, the browser may fetch multiple web app manifests and apply the appropriate one at time of use based on media queries as a way to fix this problem.

**Comment by annevk** [comment]
Can't the splash color be made to vary on a query? It seems some things in the manifest should be able to vary on this, but not all.

**Comment by marcoscaceres** [comment]
@beaufortfrancois wrote: 
> I wonder if in the future, the browser may fetch multiple web app manifests and apply the appropriate one at time of use based on media queries as a way to fix this problem.

It's probably best to just put that directly in the same manifest... similar to CSS's media rule does the various overrides. Fetching multiple manifests is going to get quite messy. 

@annevk wrote:
> As in, I would have expected that you want to make these kind of presentation decisions later, at time of use.

That would be ideal, but I'm worried that by then it's too late, because by then the browser has handed everything to the OS (and the browser might have been closed, device restarted, etc.). 

In any case, at a minimum, consistent handling of _all_ `<link rel>` things should be a good start. We can simultaneously work on adding the ability to adapt manifest properties based on media within the manifest itself.

**Comment by annevk** [comment]
I'm not sure what you mean with consistent handling, but not all attributes make sense for all HTML link types. I don't think `media` makes sense for `manifest`. (Similar to how it does not apply to `canonical`, for instance.)

**Comment by marcoscaceres** [comment]
> I'm not sure what you mean with consistent handling, but not all attributes make sense for all HTML link types.

Sorry, what I mean is that we define what those behaviors are consistently in one place - then, for example, have specific types opt-out of the affects of various attributes. So, whatever the outcome of https://github.com/whatwg/html/issues/6450 might be. 

> I don't think media makes sense for manifest. (Similar to how it does not apply to canonical, for instance.)

I don't have a strong opinion if it does or not, tbh - but I think we should discuss that and come to some agreement across all the engines.

**Comment by beaufortfrancois** [comment]
> It's probably best to just put that directly in the same manifest... similar to CSS's media rule does the various overrides. Fetching multiple manifests is going to get quite messy.

The thing I worry is that we'll have to define/update manifest properties each a new CSS media feature is added/updated.

**Comment by annevk** [comment]
@beaufortfrancois I'm not sure this is the place to get into web manifest specifics, but you could make a subset of features vary on a media query (and potentially other things), no?

**Comment by beaufortfrancois** [comment]
Ideas like as https://github.com/w3c/manifest/issues/186#issuecomment-43824192 and https://github.com/w3c/manifest/issues/955#issuecomment-786852869 are indeed some interesting ways of addressing this, but I do think using the strong power of established patterns of the web platform is **also** valuable here. Web developers are already familiar with them and it is more extensible and future-proof.

**Comment by annevk** [comment]
Sure, but patterns should only be adopted when they are applicable. As discussed in 186 they do not apply here.

**Comment by beaufortfrancois** [comment]
I do think this pattern applies as said before: a media query for light/dark mode applies for instance to web app manifests. Currently users who install the Twitter Web App for instance always get the light version, even if when they use the OS dark mode. That results in a bad user experience. This proposal addresses this issue.

**Comment by annevk** [comment]
Sure, at a high-level it does, but it doesn't address it well. You shouldn't have to generate N manifests when only a couple of fields end up being different. I think the idea in 186 that each app should have a single manifest is still valid. This could also lead to massive user confusion if app names and such end up changing based on the environment.

**Comment by beaufortfrancois** [comment]
> This could also lead to massive user confusion if app names and such end up changing based on the environment.

The manifest data is used at install time. So I'm not sure about which confusion you're talking about.

**Comment by annevk** [comment]
Fair, but then you still have the time-of-installation vs time-of-use problem and the fact that developers have to generate a bunch of manifests with nearly identical contents.

**Comment by tomayac** [comment]
As the person who initially requested this in the [Chrome bug](https://bugs.chromium.org/p/chromium/issues/detail?id=970159) blindly assuming it worked that way already, let me chime in:

I _think_ @annevk is operating under the thinking model of “whenever the media attribute changes, the manifest will be re-evaluated”, whereas my (and presumably @beaufortfrancois’) thinking model is how things currently work (that is, the manifest being re-evaluated according to https://web.dev/manifest-updates/).

I guess no browser would ever re-evaluate manifests immediately on change, but to be honest assuming they did is probably closer to user expectations, so Anne’s point. Maybe the spec should (does already?) contain non-normative text on the to-be-expected UA re-evaluation frequency. 

Also, I want to (re-)emphasize that we are not opposed to making the manifest fields media-aware as an alternative way to achieve the same with just one manifest.

As a side note, some developers probably would not actually create physical different manifests, but dynamically create them based on query parameters for example.

**Comment by annevk** [comment]
Again, even if you pick a single one and stick with it, you still have the two problems I pointed out in the comment above you. I think that's more than enough reason to not pursue this.

**Comment by marcoscaceres** [comment]
I don't think this is right place to have this discussion. We should pick this up in whatwg/html#6450.

**Comment by kenchris** [comment]
Translations is generally a big issue we should solve. 

From quick testing, I can change my language and dark mode on my Android phone, and it is reflected in apps immediately (my Twitter PWA reloaded on interaction). I guess the only way to properly support that would be to support embedded translations and maybe "media" inside the manifest file itself. But users would also need to make sure that their Service Worker cached files have meta data like language associated.

I believe @aarongustafson looked into this.

On the other hand, users are using `accept-language` and custom url's (say add `?lang=da`) today to serve different manifests depending on language (or location).

Installing Twitter here (Denmark) installs a Danish manifest, thought I am using my system in English:

![image](https://user-images.githubusercontent.com/1101339/112962786-0951e680-9147-11eb-9a84-45fec2520f2e.png)

**Comment by kenchris** [comment]
Also related: https://github.com/w3c/manifest-app-info/issues/10

**Comment by kenchris** [comment]
FYI @diekus @mgiuca

**Comment by annevk** [comment]
I'm not sure how the `media` attribute relates to languages? In any effect, I tend to agree that localization support should be possible with a single manifest as well, if you in fact have a single application across those locales.

**Comment by marcoscaceres** [comment]
> I tend to agree that localization support should be possible with a single manifest as well, if you in fact have a single application across those locales.

Agree. I'll work to refine the proposal for localization.
